src/uint/cmp.rs

```diff
@@ -25,103 +23,97 @@ impl<const LIMBS: usize> Uint<LIMBS> {
-            lt |= ((a - b) >> Limb::BITS) & 1 & !gt;
-            i -= 1;
-        }
-        (gt as SignedWord) - (lt as SignedWord)
-    }
-
-    /// Returns 0 if self == rhs or Word::MAX if self != rhs.
-    /// Const-friendly: we can't yet use `subtle` in `const fn` contexts.
-    #[inline]
-    pub(crate) const fn ct_not_eq(&self, rhs: &Self) -> Word {
+    pub(crate) const fn ct_eq(&self, rhs: &Self) -> CtChoice {
         let mut acc = 0;
         let mut i = 0;
 
         while i < LIMBS {
             acc |= self.limbs[i].0 ^ rhs.limbs[i].0;
             i += 1;
         }
-        let acc = acc as SignedWord;
-        ((acc | acc.wrapping_neg()) >> HI_BIT) as Word
+
+        // acc == 0 if and only if self == rhs
+        !Limb(acc).ct_is_nonzero()
+    }
+
+    /// Returns the truthy value if `self <= rhs` and the falsy value otherwise.
+    #[inline]
+    pub(crate) const fn ct_lt(&self, rhs: &Self) -> CtChoice {
+        // We could use the same approach as in Limb::ct_lt(),
+        // but since we have to use Uint::wrapping_sub(), which calls `sbb()`,
+        // there are no savings compared to just calling `sbb()` directly.
+        let (_res, borrow) = self.sbb(rhs, Limb::ZERO);
+        borrow.0
+    }
+
+    /// Returns the truthy value if `self <= rhs` and the falsy value otherwise.
+    #[inline]
+    pub(crate) const fn ct_gt(&self, rhs: &Self) -> CtChoice {
+        let (_res, borrow) = rhs.sbb(self, Limb::ZERO);
+        borrow.0
     }
 }
 
 impl<const LIMBS: usize> ConstantTimeEq for Uint<LIMBS> {
     #[inline]
     fn ct_eq(&self, other: &Self) -> Choice {
-        Choice::from((!self.ct_not_eq(other) as u8) & 1)
+        Choice::from(self.ct_eq(other) as u8 & 1)
     }
 }
 
 impl<const LIMBS: usize> ConstantTimeGreater for Uint<LIMBS> {
     #[inline]
     fn ct_gt(&self, other: &Self) -> Choice {
-        let underflow = other.sbb(self, Limb::ZERO).1;
-        !underflow.is_zero()
+        Choice::from(self.ct_gt(other) as u8 & 1)
     }
 }
 
 impl<const LIMBS: usize> ConstantTimeLess for Uint<LIMBS> {
     #[inline]
     fn ct_lt(&self, other: &Self) -> Choice {
-        let underflow = self.sbb(other, Limb::ZERO).1;
-        !underflow.is_zero()
+        Choice::from(self.ct_lt(other) as u8 & 1)
     }
 }
 
 impl<const LIMBS: usize> Eq for Uint<LIMBS> {}
 
 impl<const LIMBS: usize> Ord for Uint<LIMBS> {
     fn cmp(&self, other: &Self) -> Ordering {
-        match self.ct_cmp(other) {
-            -1 => Ordering::Less,
-            1 => Ordering::Greater,
-            n => {
-                debug_assert_eq!(n, 0);
-                debug_assert!(bool::from(self.ct_eq(other)));
-                Ordering::Equal
-            }
+        let is_lt = self.ct_lt(other);
+        let is_eq = self.ct_eq(other);
+
+        if is_lt == Word::MAX {
+            Ordering::Less
+        } else if is_eq == Word::MAX {
+            Ordering::Equal
+        } else {
+            Ordering::Greater
```

Comment:
Seems like you could use `subtle` here (although the mixture of `ct_eq` might get a little confusing)

```suggestion
        let is_eq = self.ct_eq(other).ct_eq(Word::MAX);
        let is_gt = self.ct_gt(other).ct_eq(Word::MAX);
        
        let mut ordering = Ordering::Less;
        ordering.conditional_assign(&Ordering::Equal, is_eq);
        ordering.conditional_assign(&Ordering::Greater, is_gt);
        ordering
```

Reply:
This method wasn't constant-time before the change, so I didn't think of that, but sure, this can be done

Reply:
After some thought, I think the current implementation is fine. This method is not supposed to be constant-time, so we probably shouldn't complicate it (plus its return value can't be used in a constant-time context anyway).

Reply:
Turns out that `Ordering` doesn't impl `ConditionallySelectable` either: https://github.com/dalek-cryptography/subtle/pull/102